ui/src/app/service/data.service.ts

```diff
@@ -104,6 +106,39 @@ export class DataService {
     });
   }
 
+  /**
+   * Add a new comment under a specified thread.
+   */
+  addComment(threadId: string, username: string, content: string, extraTags: string[]): Promise<Comment> {
+    return new Promise<Comment> ((resolve, reject) => {
+
+      const newComment = new Comment();
+      newComment.setThreadId(threadId);
+      newComment.setUsername(username);
+      newComment.setCommentContent(content);
+
+      const extraTagsList: Array<Tag> = extraTags.map(tagName  => {
+        const extraTag = new Tag();
+        extraTag.setTagName(tagName);
+        return extraTag;
+      });
+      newComment.setExtraTagsList(extraTagsList);
+
+      const req = new AddCommentUnderThreadRequest();
+      req.setComment(newComment);
+
+      this.client.addCommentUnderThread(req, {}, (err, response: AddCommentUnderThreadResponse) => {
+        if (err) {
+          console.error(err.code, err.message);
+          reject(err);
+        }
+        if (response) {
```

Comment:
If neither of `err` or `response` evaluates to `true`, the `Promise` won't be resolved.